### ISSUE TITLE
Android back button does not terminate app

### ISSUE THREAD

**Issue by yms2772** [fyne-io/fyne#1050]
### Describe the bug:

My app shows the time that app was opened using `time.Now()` so if it is terminated and opened again, the time must be changed to the current time.
But it always shows the first opening time.

Android back button does not terminate app.

### Example code:
```go
a := app.NewWithID("com.test.myapp")
a.Settings().SetTheme(NewCustomTheme())

w := a.NewWindow("Opened Time")
w.CenterOnScreen()
w.SetFixedSize(true)
w.Resize(fyne.NewSize(400, 600))

w.SetOnClosed(func() {
	os.Exit(1)
})

mainContent := widget.NewLabel(time.Now())

w.SetContent(mainContent)
w.ShowAndRun()
```

### Device (please complete the following information):
 - OS: Android
 - Version: 10
 - Go version: 1.14.3 linux/amd64
 - Fyne version: develop

**Comment by andydotxyz** [comment]
It is my understanding that an app should not be terminated when you press the back button. Android itself will control when an app terminates (like memory usage or not responding to pings).

We do need better lifecycle callbacks, but forcing apps to quit is not part of the Android behaviour (not iOS)

**Comment by yms2772** [comment]
> It is my understanding that an app should not be terminated when you press the back button. Android itself will control when an app terminates (like memory usage or not responding to pings).
> 
> We do need better lifecycle callbacks, but forcing apps to quit is not part of the Android behaviour (not iOS)

So that mean, back button cannot terminate fyne app and if I want to terminate my app, I should make a button with `os.Exit`. Is that right?

**Comment by andydotxyz** [comment]
That would not be advisable either. Some App Store review processes will reject apps that forcefully terminate.
Users can quit apps using the app switcher, that is the equivalent to closing the last window on desktop.

Can you share more info on why your app wants to exit instead of staying in the background?

**Comment by yms2772** [comment]
> That would not be advisable either. Some App Store review processes will reject apps that forcefully terminate.
> Users can quit apps using the app switcher, that is the equivalent to closing the last window on desktop.
> 
> Can you share more info on why your app wants to exit instead of staying in the background?

Because my app should display the time that it is opened.
App switcher can terminate my app, but I wanna terminate just using back button.

**Comment by andydotxyz** [comment]
In that regard it is sort of working correctly - A mobile app is opened when it first launches and then stays open until the phone reboots or the app is force-quit.

I have written up a ticket that I think captures what we could add to support some more features in this area #1053. But OnForeground would not just cover "starting" after a back button but also as a result of changing apps away then back...

**Comment by yms2772** [comment]
> In that regard it is sort of working correctly - A mobile app is opened when it first launches and then stays open until the phone reboots or the app is force-quit.
> 
> I have written up a ticket that I think captures what we could add to support some more features in this area #1053. But OnForeground would not just cover "starting" after a back button but also as a result of changing apps away then back...

I think adding some hardware events for mobile like navigation bar or volume button seem to be able to alleviate this problem.

Just like below..
```go
w.Canvas().SetOnTypedKey(func(k *fyne.KeyEvent) {
    if k.Name == fyne.KeyBackButton {
        os.Exit(0)
    }
})
```

**Comment by andydotxyz** [comment]
Such events would not be generic to mobile but specific to Android. On iOS there is no back button at all apps have been removed from the store for listening to volume press events iirc.

I don't quite understand why you are fighting the OS which is doing everything it can to avoid the delays of app startup as much as possible. I assume there must be more to your use-case than showing some metadata - if so can you expand on the importance of having a clean start to the app on every return to it? It could help inform the design of our lifecycle enhancements.

**Comment by yms2772** [comment]
> Such events would not be generic to mobile but specific to Android. On iOS there is no back button at all apps have been removed from the store for listening to volume press events iirc.
> 
> I don't quite understand why you are fighting the OS which is doing everything it can to avoid the delays of app startup as much as possible. I assume there must be more to your use-case than showing some metadata - if so can you expand on the importance of having a clean start to the app on every return to it? It could help inform the design of our lifecycle enhancements.

I'm making ToDo list from website so I wanna display ToDo list on main window.

I set the app to fetch the ToDo list from the website when it starts up so the data will not change unless the app is completely turned off.

**Comment by andydotxyz** [comment]
I see, yes data management is complex - I would not recommend loading only once. It is the job of an app to ensure its data is always up to date and using a force-quit is not really a good way to do this.

The simplest approach would be to put a “refresh” button where you were thinking about a “quit” one. This would load new data whilst the app is still running.

Better, but a bit more work, is to have the app check with the server on some regular basis whether the data has changed, and update if it has.

Best and the most work would be for the app to ask the server to notify you when data changes so that it can refresh at exactly the right time.

I hope one of these strategies helps you and avoids doing something the platforms advise against.

**Comment by elbashery** [comment]
I have a question in this context
Does the Android or iOS system allow receiving notifications from the private server?
For example, SSE services, completely cutting out GCM, Firebase, or the like ???
Similar question:
When the app is sleeping in the background, can it receive a message via socket. As mqtt, nats, or similar.
Thank you so much.

**Comment by andydotxyz** [comment]
When the app is in the foreground you should be able to connect to any servers and register for callbacks like any other Go app.

With background however the OS may suspend active connections. We need to add support for push notifications and lifecycle management to properly manage background updating.

**Comment by andydotxyz** [comment]
Closing this as won't fix, the real solution is a data handling strategy rather than forcing apps to quit so data is invalidated.
